### DISCUSSION TITLE
Can `mdxJsxFlowElement` have phrasing content as child?

### DISCUSSION THREAD

**Discussion by clemyan** [mdx-js/mdx#2571] [Q&A]
I am currently writing a markdown plugin in TypeScript to process MDX and have encountered a problem:

- If I enter `<div>Hello <span>World</span></div>` into the playground to check the mdast tree, we can see an `mdxJsxFlowElement` having `mdxJsxTextElement` and `text` as children.
- But neither the [spec](https://github.com/syntax-tree/mdast-util-mdx-jsx?tab=readme-ov-file#mdxjsxflowelement) nor the [TypeScript types](https://github.com/syntax-tree/mdast-util-mdx-jsx/blob/e0d1a16706058faf7f12815789d0c3148dfd8303/index.d.ts#L144) allows `mdxJsxTextElement` and `text` to be children of `mdxJsxFlowElement`

So which of the two is correct? Or am I missing something?

**Comment by wooorm** [comment]
Hi!

What are you doing and why? Perhaps markdown is not what you care about, perhaps you care about HTML.
I would recommend not checking for one or the other, but checking for both. TS should be smart enough to understand that content can then be both too.
Is this a problem with the playground or with actual code you have?

The spec/types you link is about the earliest parser integration.
MDX itself later does more things. Notably `remarkMarkAndUnravel` in https://github.com/mdx-js/mdx/blob/main/packages/mdx/lib/core.js is likely doing what you see.

**Reply by clemyan** [reply]
tl;dr: The MDX playground outputs an mdast tree that is not compliant to the spec.

----

I misspoke. I am making a **remark** plugin, called via Docusaurus.

The plugin, in short, reads the code blocks on the page and replace them with custom JSX elements for styling and attaching metadata. As a simplified example, imagine it transforms this:

````
```
git restore -S ./path
EDITOR=vi git commit --amend
```
````

into the something like this

```
<MyCodeBlock>
  <MyCodeLine><Command metadata={commands.git.restore}>git restore -S ./path</Command></MyCodeLine>
  <MyCodeLine><Env>EDITOR=vi</Env> <Command metadata={command.git.commit}>git commit --amend</Command></MyCodeLine>
</MyCodeBlock>
```

Doing that as a remark plugin should be simple: just use `unist-util-visit` to find all `CodeBlock` nodes in the mdast tree, and replace them with `MdxJsxFlowElement` nodes, right?

Using the MDX playground, we can see the mdast tree for the JSX snippet above is this:

```jsonc
{
  "type": "mdxJsxFlowElement",
  "name": "MyCodeBlock",
  "attributes": [],
  "children": [
    {
      "type": "mdxJsxFlowElement",
      "name": "MyCodeLine",
      "attributes": [],
      "children": [
        {
          "type": "mdxJsxTextElement",
          "name": "Command",
          "attributes": [/* --snip-- */],
          "children": [
            {
              "type": "text",
              "value": "git restore -S ./path"
            }
          ],
          "data": {
            "_mdxExplicitJsx": true
          }
        }
      ],
      "data": {
        "_mdxExplicitJsx": true
      }
    },
    {
      "type": "mdxJsxFlowElement",
      "name": "MyCodeLine",
      "attributes": [],
      "children": [
        {
          "type": "mdxJsxTextElement",
          "name": "Env",
          "attributes": [],
          "children": [
            {
              "type": "text",
              "value": "EDITOR=vi"
            }
          ],
          "data": {
            "_mdxExplicitJsx": true
          }
        },
        {
          "type": "text",
          "value": " "
        },
        {
          "type": "mdxJsxTextElement",
          "name": "Command",
          "attributes": [/* --snip-- */],
          "children": [
            {
              "type": "text",
              "value": "git commit --amend"
            }
          ],
          "data": {
            "_mdxExplicitJsx": true
          }
        }
      ],
      "data": {
        "_mdxExplicitJsx": true
      }
    }
  ],
  "data": {
    "_mdxExplicitJsx": true
  }
}
```

However, according to the spec I linked, this is not a valid mdast tree, because MdxJsxTextElement cannot be a child of MdxJsxFlowElement. And since the [TypeScript types](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBDAnmApnA3nAsgEwB4BSAzngGIA2EA7gKLkogoB28AvnAGZQQhwDkIHAEMiMALQBXGMHJjBeMQCsSfALAAoDQGMITUXCYQcKAFzZ8xMpVr1GLOAF4MGuK7cAiJKndn38yxTUdAzMMO4ANC5uru5MQow+cO5YiADCRigAQpRaANYRUdHuQjAwUMAARlIoRIkA2gC6kerRMVoAFjI4UMz1ha3o-a1JXiiJfhYkgTYhLAUtwx5xCb4p6cYAMsBMY82LHiVlldW1Zo17+0kdXT1MfQuXzg+PI8hjvv4kACooeDDBdjCFxeSWW7yS6RAICETBw8xBMUO5SqMBq9wRbkGzwxnje40+eAAgqVkdV4RilvFwX4UDAhMI6eSKUkAG5CcgScFY5lFUb4yZEknHVEANXZnJoeDAPSIRGAuiZPPcbI51J0UJhOCIADoAObAGDamUwaC7IYU9wMoSJbk8oo1MooLnmu247y+AAK3F1UHiirtMQqRkQ6IDw1tYeGbupkulNTlugAynTUYD-ZGYr847L5XczBGM61o+MsAwKigoLGZQm7sDC0UIBVFCgtGF8y76693UlSyBy5WpdXc+nO0lG83WzaO6Pi74AJLGFjADjACsj0egqnjdXQ2G1aeF1h1jdJaUQVCwEPt7EnrvUhehZerqDrmdg8b6sIHjNH7+R9zqmA1RwmYHDskQKDHhu7jntIujsokYHkBBf52r+N71u4Z4XkgU4YZ2s5JA+S4rmuUFvluvjGqa7ioTy6G3lcPBAaiIGcOBkF0cyMFgHBcTkIhHFcQirDCY8on4cyTRiYs7hEBAEhQFoKBfHiHxGBI9CvhaO6hKccCNDJ0QSWGJn0cJZkgtJklFtc5DdL0ZzCQWOJ8r4qJ-Np+zKuK1KfnA1E9HAYiJnA2oAPRgCU7S0TZwyWS81k4laeHcQA+vIsbkMAWgGpYiRlJydEJfsJWtEllyWiU1rXiC7gZfgWU5XlJAFVARU2WVrgMYsLlFm5SQEtMAKhNpsSUT2aQZFsOxjUiwpomcFXeXZDl5gZf59d5A0TIQ3y-P8tijeRsnvr4NBMCyXkHEKKKLQZy0vABnT2bcoYYltT07R5QIyT5qrjDQAAic5fAA8gASg4LLALFzJdYsj2PFVjK1RaDV4E1uUwPlZiFSgxV-j1jyfVG30HddMQqpy4xJETJ3RKT-VqYNAo-H8I1zAzRZnRCPC7nC3NFPNd36XUzl-TtBLEkcd2UzzE00nSKVC7J1POnFlVSwKMukqKvlVvGw6q956vbvzmp6ga2o7ga8tRilaNhu4Do9BrmE7V6EA+n6JtPUGOBXhtmsfUZWss+4hs5kmKazL9IcWlmQ4Kk70E7b2-ZRzW9sWuOLZtk8jExOnZYVlnxth3VeeTqnjGEe4xHSKRL5+87vMARbsJw0XxO3lh3A4UHTMexHjdPmRlcIuNKxJJ+3eMb3J4d+AwGCchnEJ66sG5ghoFCZvFKLwR2EVrhtdp6Pi5N8+Ofce3ttfgfGJH5hgGr3v6+t7nvE7wJH8oU-F4CM0KT1cEjAMckFJKRUhHEAGktJfxWvzPS9QGiT2AUAmSGD4p0XAStF6a13oImHrJcmnlEGsl8h+A0cAH7BTEFSLuFlcF+xRjVQu6NMpSmytjXGCB2oEwTtguAGC8EeEdhw5GGMsYtTwG1Dq4kOxlXAWw1KslpHcOajjVqeMBHmgSqwIAA) follows the spec, my plugin (written in TypeScript) does not compile.

My understanding is that spec (together with the unist and mdast specs) *defines* what a valid mdast tree *is* when using MDX. Am I misunderstanding something? Or is the playground outputting an invalid mdast tree? Are there TypeScript types for remark plugin authors to typecheck the plugin's output?

**Reply by wooorm** [reply]
ok, so that has little to do with markdown or HTML or how things mix.
You care solely about JS(X). You do not want markdown *inside* these things.

Use a rehype plugin. Generate an expression. Turn:

````mdx
```sh
git commit --amend
```
````

Into:

```mdx
{
  <Component>
    <Line>git commit --amend</Line>
  </Component>
}
```

Use the playground to inspect both. You want to generate that embedded `data.estree` on the expression
